pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,363 @@
+		logger.WithError(err).Warn("Failed to write discover response message")
+		return err
+	}
+	logger.Debug("Sent discover response message")
+	return nil
+}
+
+func (s *srv) handleTraffic(c echo.Context) error {
+	var latestUpstreamXTime int64
+	id := c.Param("id")
+	var auth string
+	bearerAuth := c.Request().Header.Get("Bearer")
+	if bearerAuth != "" {
+		auth = bearerAuth
+	} else {
+		auth = c.Request().Header.Get(echo.HeaderAuthorization)
+	}
+	ctx := c.Request().Context()
+	if auth != "" {
+		md := metadata.New(map[string]string{
+			"id":            id,
+			"authorization": fmt.Sprintf("Bearer %s", auth),
+		})
+		if ctxMd, ok := metadata.FromIncomingContext(s.ctx); ok {
+			md = metadata.Join(ctxMd, md)
+		}
+		ctx = metadata.NewIncomingContext(s.ctx, md)
+	}
+
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "traffic",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+
+	ctx, ids, err := s.server.FillGatewayContext(ctx, ttnpb.GatewayIdentifiers{GatewayID: id})
+	if err != nil {
+		return err
+	}
+
+	// For gateways with valid EUIs and no auth, we provide the link rights ourselves as in the udp frontend.
+	if auth == "" {
+		ctx = rights.NewContext(ctx, rights.Rights{
+			GatewayRights: map[string]*ttnpb.Rights{
+				id: {
+					Rights: []ttnpb.Right{ttnpb.RIGHT_GATEWAY_LINK},
+				},
+			},
+		})
+	}
+
+	uid := unique.ID(ctx, ids)
+	ctx = log.NewContextWithField(ctx, "gateway_uid", uid)
+
+	fp, err := s.server.GetFrequencyPlan(ctx, ids)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to get frequency plan")
+		return err
+	}
+
+	conn, err := s.server.Connect(ctx, s, ids)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to connect")
+		return err
+	}
+	if err := s.server.ClaimDownlink(ctx, ids); err != nil {
+		logger.WithError(err).Error("Failed to claim downlink")
+		return err
+	}
+	defer func() {
+		if err := s.server.UnclaimDownlink(ctx, ids); err != nil {
+			logger.WithError(err).Error("Failed to unclaim downlink")
+		}
+	}()
+
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	go func() {
+		for {
+			select {
+			case <-conn.Context().Done():
+				return
+			case down := <-conn.Down():
+				dlTime := time.Now()
+				dnmsg := messages.FromDownlinkMessage(ids, *down, int64(s.tokens.Next(down.CorrelationIDs, dlTime)), dlTime, atomic.LoadInt64(&latestUpstreamXTime))
+				msg, err := dnmsg.MarshalJSON()
+				if err != nil {
+					logger.WithError(err).Error("Failed to marshal downlink message")
+					continue
+				}
+
+				logger.Info("Send downlink message")
+				if err := ws.WriteMessage(websocket.TextMessage, msg); err != nil {
+					logger.WithError(err).Error("Failed to send downlink message")
+					conn.Disconnect(err)
+					return
+				}
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-conn.Context().Done():
+			return nil
```

Comment:
It would be nice if we can simply `return conn.Context().Err()` here and gracefully handle context cancellation by the caller